### PR TITLE
consume NoSideEffects flag in EE

### PR DESCRIPTION
**Pull request by ivanbasov**
Fixes  https://github.com/dotnet/roslyn/issues/31617

### CODE REVIEW

**Comment by cston** [comment]
@ivanbasov, this looks like the right approach.

**Comment by ivanbasov** [comment]
> @ivanbasov, this looks like the right approach.

Thank you, @cston! Added unit tests. Please review

**Comment by ivanbasov** [comment]
tagging @jinujoseph for approval

File: eng/Versions.props [hunk 1]
```diff
@@ -95,6 +95,8 @@
     <MicrosoftVisualStudioComponentModelHostVersion>16.0.198-g52de9c2988</MicrosoftVisualStudioComponentModelHostVersion>
     <MicrosoftVisualStudioCompositionVersion>15.5.23</MicrosoftVisualStudioCompositionVersion>
     <MicrosoftVisualStudioCoreUtilityVersion>16.0.203-g5d5c63964e</MicrosoftVisualStudioCoreUtilityVersion>
+    <MicrosoftVisualStudioDebuggerEngineVersion>15.9.28307</MicrosoftVisualStudioDebuggerEngineVersion>
+    <MicrosoftVisualStudioDebuggerMetadataVersion>15.9.28307</MicrosoftVisualStudioDebuggerMetadataVersion>
```
Comment (tmat) [inline eng/Versions.props]: These are not used. We need to update `MicrosoftVisualStudioDebuggerEngineimplementationVersion` and `MicrosoftVisualStudioDebuggerMetadataimplementationVersion` instead.

Comment (tmat) [reply eng/Versions.props]: I wonder why do the test pass :-/

Comment (tmat) [reply eng/Versions.props]: The packages are in https://www.myget.org/F/roslyn_concord/api/v3/index.json, but the last version there is 15.7.

**Pull request merged**